### PR TITLE
POC: QemuHCK: Load coresponding IOMMU device

### PR DESCRIPTION
**Pull request by kostyanf14**

### CODE REVIEW

**Comment by kostyanf14** [comment]
@Jedoku 

As you have AMD CPU can you please test the PR

File: lib/setupmanagers/qemuhck/devices/amd-iommu.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "iommu",
+  "command_line": [
+    "-device amd-iommu,intremap=on,device-iotlb=on,xtsup=on"
```
Comment (Jedoku) [inline lib/setupmanagers/qemuhck/devices/amd-iommu.json:4]: `CmdRun PID:59231: stderr: qemu-system-x86_64: -device amd-iommu,intremap=on,device-iotlb=on,xtsup=on: Property 'amd-iommu.xtsup' not found`

Comment (Jedoku) [reply lib/setupmanagers/qemuhck/devices/amd-iommu.json:4]: It works well without `xtsup=on`

Comment (kostyanf14) [reply lib/setupmanagers/qemuhck/devices/amd-iommu.json:4]: This is a new flag that was added by the AMD team. Thanks for testing.

File: lib/setupmanagers/qemuhck/devices/iommu.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  class IommuDevice
+    extend T::Sig
+
+    def initialize(logger)
+      @logger = logger
+    end
+
+    sig { returns(Models::QemuHCKDevice) }
+    def qemu_hck_device
+      vendor = Sys::CPU.processors[0].vendor_id
```
Comment (akihikodaki) [inline lib/setupmanagers/qemuhck/devices/iommu.rb:16]: The gem is too complicated just to get the vendor ID. You can do: `File.read('/proc/cpuinfo').match(/^vendor_id\s*:\s*(.*)/)[1]`

File: lib/setupmanagers/qemuhck/devices/iommu.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+require 'sys/cpu'
+
+module AutoHCK
+  class IommuDevice
+    extend T::Sig
+
+    def initialize(logger)
+      @logger = logger
+    end
```
Comment (akihikodaki) [inline lib/setupmanagers/qemuhck/devices/iommu.rb:12]: I think it is simpler to have a `module` named `QemuHCK::Devices` and have a class method named `iommu`. This removes the need to assign an instance variable.

**Review by akihikodaki — CHANGES_REQUESTED** [review]

**Comment by akihikodaki** [comment]
> @akihikodaki
> 
> What do you think about this implementation of dynamic IOMMU for AMD/Intel? Do you have other suggestions?

I think the idea of having a "dynamic" device is good. I wrote a few comments on implementation details.

File: lib/setupmanagers/qemuhck/devices/iommu.rb (hunk 3 above)
Comment (kostyanf14) [reply lib/setupmanagers/qemuhck/devices/iommu.rb:12]: Do you expect to call `device.public_send("iommu")` or have a list of condition `if dev == 'iommu' then device.iommu` ?

Comment (akihikodaki) [reply lib/setupmanagers/qemuhck/devices/iommu.rb:12]: You can still keep the logic you already have in `read_dynamic_device`. Just replace `AutoHCK.const_get` with `QemuHCK::Devices.public_send`.

**Comment by kostyanf14** [comment]
Moved to https://github.com/HCK-CI/AutoHCK/pull/469

**Pull request closed without merging**